homeassistant/helpers/entityfilter.py

```diff
@@ -30,31 +39,121 @@ def convert_filter(config: Dict[str, List[str]]) -> Callable[[str], bool]:
         vol.Optional(CONF_EXCLUDE_DOMAINS, default=[]): vol.All(
             cv.ensure_list, [cv.string]
         ),
+        vol.Optional(CONF_EXCLUDE_ENTITY_GLOBS, default=[]): vol.All(
+            cv.ensure_list, [cv.string]
+        ),
         vol.Optional(CONF_EXCLUDE_ENTITIES, default=[]): cv.entity_ids,
         vol.Optional(CONF_INCLUDE_DOMAINS, default=[]): vol.All(
             cv.ensure_list, [cv.string]
         ),
+        vol.Optional(CONF_INCLUDE_ENTITY_GLOBS, default=[]): vol.All(
+            cv.ensure_list, [cv.string]
+        ),
         vol.Optional(CONF_INCLUDE_ENTITIES, default=[]): cv.entity_ids,
     }
 )
 
 FILTER_SCHEMA = vol.All(BASE_FILTER_SCHEMA, convert_filter)
 
 
+def convert_filter_alt(
+    config: Dict[str, Dict[str, List[str]]]
+) -> Callable[[str], bool]:
+    """Convert the alternate filter schema into a filter."""
+    include = config.get(CONF_INCLUDE, {})
+    exclude = config.get(CONF_EXCLUDE, {})
+    filt = generate_filter(
+        include.get(CONF_DOMAINS, []),
+        include.get(CONF_ENTITIES, []),
+        exclude.get(CONF_DOMAINS, []),
+        exclude.get(CONF_ENTITIES, []),
+        include.get(CONF_ENTITY_GLOBS, []),
+        exclude.get(CONF_ENTITY_GLOBS, []),
```

Comment:
It looks like these are always passing though the schema and will always get defaults set, we should not use `.get`

`include = config[CONF_INCLUDE]`

Reply:
The issue here is the existing tests. At runtime, there's no issue because when the schema is validated it has defaults for all those fields. So at runtime there's never a case when this function will be handed a config that does not have a `CONF_INCLUDE` OR `CONF_EXCLUDE` field or `CONF_DOMAINS`, `CONF_ENTITIES`, and `CONF_ENTITY_GLOB` fields within each of those.

But there are existing tests that assume the config is still valid even without those fields present. Specifically there are these 5 in `logbook` which do this by calling `entities_filter = logbook._generate_filter_from_config({})`:
- https://github.com/mdegat01-1/EntityFilterGLOB/blob/glob-filter-support/tests/components/logbook/test_init.py#L157
- https://github.com/mdegat01-1/EntityFilterGLOB/blob/glob-filter-support/tests/components/logbook/test_init.py#L180
- https://github.com/mdegat01-1/EntityFilterGLOB/blob/glob-filter-support/tests/components/logbook/test_init.py#L216
- https://github.com/mdegat01-1/EntityFilterGLOB/blob/glob-filter-support/tests/components/logbook/test_init.py#L249
- https://github.com/mdegat01-1/EntityFilterGLOB/blob/glob-filter-support/tests/components/logbook/test_init.py#L702

There's also one in `recorder` which does this but that one is one I had to adjust so I can fix that.

Since there were tests assuming the empty object was valid config in tests I choose to use `get` and keep those tests passing rather then adjust the tests. I can change those tests instead if we think that is the better approach.

Reply:
> edit: It looks like this use case doesn't always pass though the schema, only the other case does

This function is brand new so the only place it does this is in `logbook` in `_generate_filter_from_config`. So I think you're right, we should fix this, especially if we're going to fix `convert_filter_alt`. I think we have 3 options:

1. Adjust `_generate_filter_from_config` to just call `generate_filter` directly and do the defaulting itself as necessary.
2. Adjust the tests that bank on being able to pass `{}` as valid config to instead pass `ALT_BASE_FILTER_SCHEMA({})` to ensure defaults are set in tests too.
3. Refactor `logbook` and tests to eliminate references to `_generate_filter_from_config`. It's not really necessary anymore since the filter is built from the config as part of schema validation now.

What do you think?

Reply:
> 
> 1. Adjust `_generate_filter_from_config` to just call `generate_filter` directly and do the defaulting itself as necessary.

Ideally we remove `_generate_filter_from_config` from `logbook` and call `convert_filter_alt` (or whatever the name ends up being) in `_get_events`.  

> 2. Adjust the tests that bank on being able to pass `{}` as valid config to instead pass `ALT_BASE_FILTER_SCHEMA({})` to ensure defaults are set in tests too.

The test aren't sufficiently reflecting the input that is actually received so the tests should be adjusted.  In short our mocking isn't good enough in the tests because it is missing the schema validation.

> 3. Refactor `logbook` and tests to eliminate references to `_generate_filter_from_config`. It's not really necessary anymore since the filter is built from the config as part of schema validation now.

See above.
> 
> What do you think?

Reply:
Oh right I forgot I couldn't built the entity filter during schema validation for logbook as it created a whole host of other issues, has to be built in `_get_events`. Ok will make this adjustment.

Reply:
> Oh right I forgot I couldn't built the entity filter during schema validation for logbook as it created a whole host of other issues, has to be built in `_get_events`. Ok will make this adjustment.

It looks like we are rebuilding the filter function every time we call the logbook api for a filter that never changes 🤦    We can fix that in a followup though (just build it once in `async_setup`)